text/projects/Krzysztof_Pióro/Maximal_Suffix.tex

```diff
@@ -0,0 +1,138 @@
+                \State $j := j + k + 1$;
+            \EndIf
+            \If {$i = j}$
+                \State $j := j + 1$;
+            \EndIf
+        \EndWhile
+        \State \Return $i$
+    \end{algorithmic}
+\end{algorithm}
+
+
+Algorytm oczywiście działa w stałej (dodatkowej) pamięci i w liniowej liczbie operacji.
+
+\section{Dowód poprawności}
+
+Niech $(i, j) \rightarrow (i', j')$ oznacza, że z konfiguracji $(i, j)$  w jednej iteracji przechodzimy do $(i', j')$ i niech $\rightarrow^{*}$ będzie
+domknięciem przechodnim relacji $\rightarrow$. 
+
+Pokażemy, że po każdej głównej iteracji będą zachodziły poniższe \textbf{niezmienniki} (gdzie oznaczamy $u := w[i\ldots j-1]$):
+\begin{enumerate}
+    \item $(i < j < n) \Rightarrow u$ jest \textit{samo-maksymalne} oraz $period(u) = u$
+    \item maksymalny sufiks słowa $w$ nie zaczyna się przed $i$
+\end{enumerate}
+
+
+
+\begin{proof}[\textbf{Dowód niezmienników}]
+Początkowo $i = 1, j = 2$ i niezmiennik jest spełniony
+
+Rozważmy iterację, w której $i$ jest zmienione po raz pierwszy. Pokażemy, że niezmienniki są spełnione przed tą iteracją 
+(autorzy pracy piszą tutaj tylko, że jest to "easy to see", ale my to uzasadnimy). Oczywiście drugi niezmiennik jest spełniony, zatem skupimy się na pierwszym.
+
+Rozważamy zatem przejście $(i=1, j) \rightarrow (i=1, j' = j + k + 1)$, gdzie niezmienniki zachodzą dla słowa $u = w[i \ldots j-1]$.
+
+Wtedy niech $u' := w[i\ldots j + k] = u^tva$, gdzie $v$ jest ścisłym prefiksem $u$ (być może pustym), a $a$ jest literką taką, że $va < u$.
+
+Z \textit{samo-maksymalności} słowa $u$ oraz faktu, że $va < u$ możemy wywnioskować, że słowo $u'$ również jest \textit{samo-maksymalne}.
+Podobnie możemy też wywnioskować, że słowo $u^tv$ jest \textit{samo-maksymalne} (skorzystamy z tego później). 
+
+Aby udowodnić, że $period(u') = u'$ pokażemy, że $u'$ nie ma żadnego właściwego prefikso-sufiksu. 
+Każdy prefikso-sufiks $u'$ musi być przedłużeniem prefikso-sufiksu $w[i\ldots j+k-1]$. 
+Najdłuższy prefikso-sufiks słowa $w[i\ldots j+k-1]$ to prefiks $p = w[i\ldots i+k-1]$ (odpowiada mu sufiks $s = w[j\ldots j+k-1] = u^{t-1}v$).
+Z działania algorytmu wynika, że po prefiksie $p$ występuje litera $b$ taka, że $b > a$. 
+Zauważmy dodatkowo, że wszystkie literki występujące po prefikso-sufiksach słowa $p$ są większe bądź równe $b$ (wpp. mielibyśmy sprzeczność z \textit{samo-maksymalnością $u^tv$)}.
+Zatem nie istnieje prefikso-sufiks słowa $w[i\ldots j+k-1]$, którego dałoby się rozszerzyć do prefikso-sufiksu słowa $u'$, czyli $period(u') = u'$.
+
+\vspace{5pt}
+
+\textbf{Pierwsza modyfikacja zmiennej i:} Rozpatrzmy zatem pierwszą zmianę zmiennej $i$ z $i=0$ na $i' = i + k + 1$. 
+
+Wtedy $w[i\ldots j+k] = u^t vb$, gdzie $v$ jest ścisłym prefiksem $u$ oraz $u < vb$.\\
+Niech $m$ będzie pierwszą pozycją za $u^t$. Pokażemy, że (częściowa) historia algorytmu wygląda jak poniżej:
+\begin{align*}
+    (i, j) \rightarrow (i', j) \rightarrow^{*} (m, j) \rightarrow^{*}(m, m+1)
```

Comment:
1. Probably a silly question but is it possible that $i \ge j$? And if so, does this imply that $u$ is empty? (which case problems for $v$ since it has to be a proper prefix of $u$)
2. "(m, m+1)" - the first bracket got mis-formatted, see pdf

Reply:
Regarding point 2. I don't see an issue here. I think it is possible that you were looking at old version of pdf (I've done some updates to docs 2 days after I've created PR). Could you confirm that?

Reply:
Regarding point 1. In that part we are only considering first modification of variable $i$ so we have $i < j$ and both $u$ and $v$ can be properly defined. 

In whole algorithm it is possible that $i \geq j$, but that also isn't a problem for us, because the first invariant has assumption that $i < j \leq n$.